<a href="https://colab.research.google.com/github/AsthanaShivanshi/2026_Intro_Python/blob/main/live/part-I/1.2-data-structures-and-control-flow-exercises.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

> Notebook-friendly copy of `part-I/1.2-data-structures-and-control-flow-exercises.ipynb`, generated by `tools/make_live.py`. Edit the book notebook, not this file.

In [253]:
# --- environment setup (generated, not part of the lesson) ---
# Colab and Kaggle do not ship every package this notebook imports.
# This is a no-op in an environment that is already set up.
import importlib.util
import subprocess
import sys

for module, package in {"pooch": "pooch"}.items():
    if importlib.util.find_spec(module) is None:
        subprocess.run([sys.executable, "-m", "pip", "install", "-q", package], check=True)

# Exercises

**ℹ️ How to use these exercises**

Attempt each from scratch in the empty code cell. Write plain functions with clear names, and avoid a mutable default argument.

## Exercise 1: Lists — build, modify, slice

Start from the daily mean temperatures `[-2.3, -1.1, 0.4, 1.2, -0.8]` (°C).

1. Append a sixth reading, `-3.1`.
2. Remove the value `0.4` (the sensor flagged it as unreliable).
3. Print the first reading, the last reading, and a slice containing the second and third.
4. Print how many readings remain.

In [254]:
daily_means= [-2.3, -1.1, 0.4, 1.2, -0.8]

daily_means.append(-3.1)

In [255]:
print(daily_means)

[-2.3, -1.1, 0.4, 1.2, -0.8, -3.1]


In [256]:
daily_means.remove(0.4)

In [257]:
#Removing 0.4
print(daily_means[0])
print(daily_means[-1])
print(daily_means[1:2])

-2.3
-3.1
[-1.1]


In [258]:
print (len(daily_means)) #prints the entire length

5


## Exercise 2: Dictionaries — build, look up, remove

You are given station records as a list of tuples `(code, name, elevation_m)`:

```python
[("JFJ", "Jungfraujoch", 3571), ("BAS", "Basel-Binningen", 316), ("LUG", "Lugano", 273)]
```

1. Build a dict mapping each `code` to its `elevation_m`.
2. Print the elevation for `"JFJ"`.
3. Look up `"ZRH"` with `.get`, returning `"unknown"` if it is absent.
4. Delete the entry for `"LUG"`, then print each remaining code and its elevation using
   `.items()`.

In [259]:
dict= {"JFJ" : 3571,
       "BAS": 316,
       "LUG": 273}

print(f"Elevation for JFJ is {dict['JFJ']} metres")

Elevation for JFJ is 3571 metres


In [260]:
dict.get("ZRH", "unknown") #Dictionary lookup

'unknown'

In [261]:
dict.get("JFJ") #Key lookup for JFJ,, returns elevation

3571

Alternative way

In [262]:
if "ZRH" not in dict.keys():
     print("unknown")

unknown


## Exercise 3: A pure function

Write a function `mean(values)` (no type hints) that returns the average of a list of numbers. Test it on `[6.1, 7.4, 5.9, 8.2, 6.8]`, printing the result rounded to two decimals.

In [263]:
numbers_list= [6.1, 7.4, 5.9, 8.2, 6.8]


In [264]:
def mean(values):
    total=0
    for value in values:
        total += value
    return total / len(values)

In [265]:
print(f"Mean: {mean(numbers_list):.2f}")

Mean: 6.88


## Exercise 4: A function with a default argument

Write a function `is_warm(temp_celsius, threshold=10.0)` (no type hints) that returns `True` when the temperature is at or above the threshold and `False` otherwise, using `if`/`else`. Test it on 6.1 °C with the default threshold, and again on 6.1 °C with a threshold of 5.0 °C passed by keyword.

In [266]:
def is_warm(temp_degC, threshold=10):
    if temp_degC >= threshold:
        return True
    else:
        return False

In [267]:
is_warm(6.1) #Uses default args if nothing specified. 

False

In [268]:
is_warm(6.1, threshold=5)

True

## Exercise 5: Pair two stations with zip

Two stations recorded the same five days:

```python
dates = ["Mon", "Tue", "Wed", "Thu", "Fri"]
jfj_celsius = [-2.3, -1.1, 0.4, 1.2, -0.8]
bas_celsius = [6.1, 7.4, 5.9, 8.2, 6.8]
```

Using `zip`, print one line per day showing the date and the temperature difference
(Basel minus Jungfraujoch), rounded to two decimals.

For each dict,, items,, keys and values.. 

In [269]:
dates = ["Mon", "Tue", "Wed", "Thu", "Fri"]
jfj_celsius = [-2.3, -1.1, 0.4, 1.2, -0.8]
bas_celsius = [6.1, 7.4, 5.9, 8.2, 6.8]

In [270]:
for date, jfj, bas in zip(dates, jfj_celsius, bas_celsius):
    print(f"The temperature diff between basel and jfj on {date} is {(bas - jfj):.2f}°C")

The temperature diff between basel and jfj on Mon is 8.40°C
The temperature diff between basel and jfj on Tue is 8.50°C
The temperature diff between basel and jfj on Wed is 5.50°C
The temperature diff between basel and jfj on Thu is 7.00°C
The temperature diff between basel and jfj on Fri is 7.60°C


## Exercise 6: Number the readings with enumerate

A station wrote one reading per day for a week:

```python
week_celsius = [6.1, 7.4, 5.9, 8.2, 11.3, 9.8, 6.8]
```

1. Using `enumerate`, print one line per reading: the day number and the temperature. Count days
   from 1, not 0 — `enumerate` starts at 0, so the day number is the index plus one.
2. In the same loop, keep track of the warmest reading seen so far and the day it fell on, in two
   variables defined before the loop. Do not use `max`.
3. After the loop, print the warmest reading and its day number.

In [271]:
week_celsius = [6.1, 7.4, 5.9, 8.2, 11.3, 9.8, 6.8]

In [272]:
for day_number, temperature in enumerate(week_celsius):
    print(f"day {day_number+1}: {temperature} degC")

day 1: 6.1 degC
day 2: 7.4 degC
day 3: 5.9 degC
day 4: 8.2 degC
day 5: 11.3 degC
day 6: 9.8 degC
day 7: 6.8 degC


In [273]:
def warmest_day(week_celsius):
    warmest_day=0
    for i in range(1, len(week_celsius)):
        if week_celsius[i] > week_celsius[warmest_day]:
            warmest_day = i
    return warmest_day

In [274]:
print(f"Warmest day is day {warmest_day(week_celsius)+1} with a temperature of {week_celsius[warmest_day(week_celsius)]} degC")

Warmest day is day 5 with a temperature of 11.3 degC


## Exercise 7: while, break, and continue

A logger returned six readings, two of which are missing:

```python
mixed_celsius = [6.1, None, 5.9, None, 9.5, 6.8]
```

Using a `while` loop over the indices:

1. Skip any missing value with `continue`.
2. Stop at the first reading above 9.0 °C with `break`, storing it in `first_hot`.
3. Print `first_hot` after the loop. Define it before the loop so the code is safe even if no
   reading exceeds the threshold.

In [275]:
mixed_celsius = [6.1, None, 5.9, None, 9.5, 6.8]

In [276]:
first_hot = None
for i in range(len(mixed_celsius)):
    if mixed_celsius[i] is None:
        continue
    if mixed_celsius[i] > 9:
        first_hot = mixed_celsius[i]
        break

In [277]:
print(first_hot)

9.5


## Exercise 8: Default arguments and returning two values

Write a function `summarise(values, scale=1.0)` that multiplies every value by `scale` and
returns **two** things: the total and the mean.

1. Call it on `[6.1, 7.4, 5.9, 8.2, 6.8]` with the default scale, unpacking the result into
   `total_celsius` and `mean_celsius`.
2. Call it again with `scale=2.0` passed by keyword, and print only the mean.
3. Call it once more and store the result in a single name. Print its `type` — what did you get?

In [278]:
def summarise(values, scale=1.0):
    for i in range(len(values)):
        values[i] *= scale
        total = sum(values)
        average = sum(values) / len(values)

    return total, average


In [279]:
celsius_values= [6.1, 7.4, 5.9, 8.2, 6.8]

In [280]:
total_celsius, mean_celsius = summarise(celsius_values)

In [281]:
print(total_celsius, mean_celsius)

34.4 6.88


In [282]:
total_celsius_scale_2, mean_celsius_scale_2 = summarise(celsius_values, scale=2.0)

In [283]:
print(mean_celsius_scale_2)

13.76


In [284]:
summaries= summarise(celsius_values, scale=2.0) #Unpacking values into a single tuple

In [285]:
type(summaries)


tuple

In [286]:
summaries

(137.6, 27.52)

## Exercise 9: A variable number of arguments

Write a function `warmest(*temps_celsius)` (no type hints) that returns the highest value among
any number of positional arguments, using a loop and a running maximum (not `max`). Test it on
three individual readings passed directly, and again by unpacking the list
`[6.1, 7.4, 5.9, 8.2, 6.8]` with `*`.

In [287]:
def warmest(*temps_celsius): #Can have any number of temperature arguments
    max=0
    for temp in temps_celsius:
        if temp > max:
            max = temp
    return max

In [288]:
list_temps=[6.1, 7.4, 5.9, 8.2, 6.8]
warmest_temp= warmest(*list_temps)
print(warmest_temp)

8.2


## Exercise 10: Fix the shared-default bug

An assistant wrote this function; it is correct the first time but wrong on later calls:

```python
def collect_above(values, threshold, out=[]):
    for v in values:
        if v > threshold:
            out.append(v)
    return out
```

In one comment, explain the bug, then rewrite it so each call starts fresh. Demonstrate with two
separate calls, assigning each result to its own variable.

In [289]:
def collect_above(values, threshold, out=None):
    for v in values:
        if v > threshold:
            out.append(v)
    return out

In [290]:
list_values=[1, 5, 8, 3, 10]

In [291]:
collected=collect_above(list_values, 5, out=[])
print(collected)

[8, 10]


## Going deeper (optional)

### Exercise 11: Document your function with a docstring

Take the `mean(values)` you wrote in Exercise 3 and give it a docstring — a one-line summary in
triple quotes on the first line of the body. State what it returns *and* the unit it assumes, so
the function carries information the name alone cannot.

Then print `mean.__doc__`, and call `help(mean)` to see what a reader of your code would see.

In [292]:
def warmest(*temps_celsius): #Can have any number of temperature arguments

    """Returns warmest temperature from the given arguments"""
    max=0
    for temp in temps_celsius:
        if temp > max:
            max = temp
    return max

In [293]:
warmest.__doc__

'Returns warmest temperature from the given arguments'

In [294]:
help(warmest)

Help on function warmest in module __main__:

warmest(*temps_celsius)
    Returns warmest temperature from the given arguments



### Exercise 12: Tuples — unpacking and immutability

A station's position is stored as a tuple, `station_coords = (46.5475, 7.9853)`.

1. Unpack it into `lat` and `lon` in one line, and print both.
2. Try to change the latitude to `47.0`. Catch the resulting error with `try`/`except TypeError`
   and print the message.
3. In a comment, say why a coordinate pair is better stored as a tuple than as a list.

In [295]:
station_coords = (46.5475, 7.9853)

In [296]:
lat,lon = station_coords

In [297]:
print(lat, lon)

46.5475 7.9853


In [298]:
print(station_coords[0], station_coords[1])

46.5475 7.9853


In [299]:
station_coords[0]=47

TypeError: 'tuple' object does not support item assignment

In [ ]:
#So that you dont change the coordinates and dates and stuff like this by mistake in your dataset. A coordinate should remain immutable

In [ ]:
#Howevrt, if you unpack it
lat,lon = station_coords
print(lat, lon)
lat=47
lon=station_coords[1]

46.5475 7.9853


In [ ]:
print(lat, lon) #It did change once you unpack it

47 7.9853


### Exercise 13: Means and a filter, with comprehensions

Recreate a `stations` dict where each code maps to a dict containing `readings_celsius`:

```python
{"JFJ": {"readings_celsius": [-2.3, -1.1, 0.4, 1.2, -0.8]},
 "BAS": {"readings_celsius": [6.1, 7.4, 5.9, 8.2, 6.8]},
 "LUG": {"readings_celsius": [9.4, 10.2, 8.8, 11.1, 9.9]}}
```

With a dict comprehension, build `means = {code: mean_temp}`, calling the `mean` function you
wrote in Exercise 3 rather than summing inline. Then, with a list comprehension, build the list
of codes whose mean is below 0 °C.

In [ ]:
stations= {"JFJ": {"readings_celsius": [-2.3, -1.1, 0.4, 1.2, -0.8]},
 "BAS": {"readings_celsius": [6.1, 7.4, 5.9, 8.2, 6.8]},
 "LUG": {"readings_celsius": [9.4, 10.2, 8.8, 11.1, 9.9]}}

In [ ]:
#Repasting it here 

def mean(values):
    total=0
    for value in values:
        total += value
    return total / len(values)

In [ ]:
for code, dictionary in stations.items():
    mean_temps = mean(dictionary['readings_celsius'])
    means[code] = mean_temps

print(means)

{'LUG': 9.879999999999999, 'JFJ': -0.52, 'BAS': 6.88}


### Exercise 14: Spot the impure function

This function is meant to return the readings without the last one:

```python
def drop_last(readings):
    readings.pop()
    return readings
```

1. Call it on `original = [6.1, 7.4, 5.9]`, then print `original`. What happened?
2. Rewrite it as a pure function that leaves the caller's list untouched, and demonstrate that
   `original` is unchanged after the call.
3. In one comment, connect this to the mutable-default bug from the lecture.

In [ ]:
def drop_last(readings):
    copy_readings= readings.copy() #Create the copy
    copy_readings.pop()
    return copy_readings

In [ ]:
original = [6.1, 7.4, 5.9]
modified = drop_last(original)
print(modified)
print(original)

[6.1, 7.4]
[6.1, 7.4, 5.9]


### Exercise 15: Reject impossible temperatures

Write `to_kelvin(temp_celsius)` that raises a `ValueError` when the input is below absolute zero
(−273.15 °C) and otherwise returns the temperature in kelvin. Give it a docstring and a type hint.

Loop over `[20.0, -300.0, 0.0]`, calling it inside a `try`/`except` so that the invalid value is
reported without stopping the loop.

In [ ]:
def to_kelvin(temp_celsius:float) -> float:
    if temp_celsius < -273.15:
        raise ValueError("This temp is impossible")
    return temp_celsius + 273.15


In [ ]:
temp_list= [20.0, -300.0, 0.0]

In [ ]:
for temp in temp_list:
    try:
        print(to_kelvin(temp))
    except ValueError as e:
        print(e)

293.15
This temp is impossible
273.15


## Exercise 16: A year of station data, continued

The optional section ends with Exercise 15. Exercises 16 and 17 are the main exercises of this
subchapter.

In subchapter 1.1 you fetched this file, worked out its structure, and converted a single
record to °C. You did not yet have the tools to handle all of it at once — now you do.

The goal: turn a year of daily maximum temperatures into a monthly summary, and write that
summary out as a new file. The input file has the same three columns as in Exercise 9 of the
previous subchapter: the station code, the day as `dd.mm.yy`, and the daily maximum temperature
in degrees Fahrenheit. Your monthly summary will have its own columns: `month`, `n_obs`
(the number of observations), and `mean_temp_celsius`.

In [ ]:
# Pre-supplied: download the data file and cache it locally.
# You do not need to understand this cell yet — fetching data is covered in the
# reproducible-data-pipelines bonus subchapter.
import pooch

data_file = pooch.retrieve(
    url="https://raw.githubusercontent.com/gse-unil/2026_MLEES_book/main/data/part-I/station_iib_daily_max_temp_2022.csv",
    known_hash="sha256:034755fb289b4e157e5d029995481786a359eb25a80df62c09ee83d063013144",
    fname="station_iib_daily_max_temp_2022.csv",
    path=pooch.os_cache("mlees"),
)
print("data cached at:", data_file)

data cached at: /Users/sasthana/Library/Caches/mlees/station_iib_daily_max_temp_2022.csv


**Step 1.** Write a function `fahrenheit_to_celsius(temp_fahrenheit)` that returns the
temperature in °C. Test it on the freezing point of water (32 °F should give 0.0 °C) and on
212 °F. A one-line docstring is optional here (see Exercise 11).

You wrote this conversion inline in 1.1. Steps 2 and 4 both need it again; with a function,
they call the same tested code instead of repeating the arithmetic.

In [ ]:
def fahrenheit_to_celsius(temp_fahrenheit: float) -> float:
    return (temp_fahrenheit - 32) * 5/9

In [ ]:
fp_water= fahrenheit_to_celsius(32)
print(fp_water)

bp_water= fahrenheit_to_celsius(212)
print(bp_water)

0.0
100.0


**Step 2.** Open the file, read all lines, and loop over the records (skipping the header).
For each one, extract the month and the temperature in °C. Print the first three as a check.

Sixteen days in this file have an empty temperature field — a real gap in the station's record,
late October to mid-November. `float("")` raises `ValueError`, so your loop has to decide what to
do with those days before it can get to the end of the file. Skip them, and count how many you
skipped. After the loop, print how many records you converted and how many you skipped; the two
should add up to 365.

In [ ]:
from pathlib import Path

data_path = Path("/Users/sasthana/Library/Caches/mlees/station_iib_daily_max_temp_2022.csv")

In [ ]:
with data_path.open("r") as f:

    lines = f.read().splitlines()
    

In [ ]:
lines= lines[1:] #Skipping header
print(lines)

['IIB,01.01.22,14.7', 'IIB,02.01.22,8.6', 'IIB,03.01.22,24.4', 'IIB,04.01.22,33.8', 'IIB,05.01.22,21.6', 'IIB,06.01.22,1.4', 'IIB,07.01.22,8.6', 'IIB,08.01.22,30.9', 'IIB,09.01.22,24.6', 'IIB,10.01.22,14', 'IIB,11.01.22,39.4', 'IIB,12.01.22,41.9', 'IIB,13.01.22,43', 'IIB,14.01.22,34.9', 'IIB,15.01.22,18.5', 'IIB,16.01.22,28.9', 'IIB,17.01.22,29.1', 'IIB,18.01.22,38.3', 'IIB,19.01.22,21', 'IIB,20.01.22,4.3', 'IIB,21.01.22,15.6', 'IIB,22.01.22,24.4', 'IIB,23.01.22,19.4', 'IIB,24.01.22,26.1', 'IIB,25.01.22,4.1', 'IIB,26.01.22,17.6', 'IIB,27.01.22,35.2', 'IIB,28.01.22,14.9', 'IIB,29.01.22,29.8', 'IIB,30.01.22,24.8', 'IIB,31.01.22,55', 'IIB,01.02.22,37.9', 'IIB,02.02.22,12.9', 'IIB,03.02.22,11.5', 'IIB,04.02.22,16', 'IIB,05.02.22,28.4', 'IIB,06.02.22,36.7', 'IIB,07.02.22,22.6', 'IIB,08.02.22,41.5', 'IIB,09.02.22,39.4', 'IIB,10.02.22,38.7', 'IIB,11.02.22,39.9', 'IIB,12.02.22,13.8', 'IIB,13.02.22,12.2', 'IIB,14.02.22,28.2', 'IIB,15.02.22,41.5', 'IIB,16.02.22,43.7', 'IIB,17.02.22,27.5', 'IIB,1

In [ ]:
count = {}
totals = {}
skipped = 0

In [301]:
for line in lines:
    
    fields = line.strip().split(",")


    if len(fields) != 3:
        skipped += 1 
        continue


    _,date,temp = fields

    if temp == "":
        skipped += 1
        continue



    try:
        temp = float(temp)

    except ValueError:
        skipped += 1
        continue



    day= date.split(".")[0]


    months = date.split(".")[1]

    #print(fields)
    #print(dates)
    #print(months)
    #print(days)
    #print(temps)


    print(f"Day: {day}, Temp: {temp}")


    print(f"Skipped: {skipped}")


            

Day: 01, Temp: 14.7
Skipped: 48
Day: 02, Temp: 8.6
Skipped: 48
Day: 03, Temp: 24.4
Skipped: 48
Day: 04, Temp: 33.8
Skipped: 48
Day: 05, Temp: 21.6
Skipped: 48
Day: 06, Temp: 1.4
Skipped: 48
Day: 07, Temp: 8.6
Skipped: 48
Day: 08, Temp: 30.9
Skipped: 48
Day: 09, Temp: 24.6
Skipped: 48
Day: 10, Temp: 14.0
Skipped: 48
Day: 11, Temp: 39.4
Skipped: 48
Day: 12, Temp: 41.9
Skipped: 48
Day: 13, Temp: 43.0
Skipped: 48
Day: 14, Temp: 34.9
Skipped: 48
Day: 15, Temp: 18.5
Skipped: 48
Day: 16, Temp: 28.9
Skipped: 48
Day: 17, Temp: 29.1
Skipped: 48
Day: 18, Temp: 38.3
Skipped: 48
Day: 19, Temp: 21.0
Skipped: 48
Day: 20, Temp: 4.3
Skipped: 48
Day: 21, Temp: 15.6
Skipped: 48
Day: 22, Temp: 24.4
Skipped: 48
Day: 23, Temp: 19.4
Skipped: 48
Day: 24, Temp: 26.1
Skipped: 48
Day: 25, Temp: 4.1
Skipped: 48
Day: 26, Temp: 17.6
Skipped: 48
Day: 27, Temp: 35.2
Skipped: 48
Day: 28, Temp: 14.9
Skipped: 48
Day: 29, Temp: 29.8
Skipped: 48
Day: 30, Temp: 24.8
Skipped: 48
Day: 31, Temp: 55.0
Skipped: 48
Day: 01, Temp

In [ ]:
# Your solution here
# Hint: open it as in 1.1, with Path(data_file).open("r", encoding="utf-8") in a with block
# Hint: lines[1:] skips the header row
# Hint: .strip().split(",") gives you the three fields
# Hint: the month is the middle piece of the day field — split it on "."
# Hint: use a counter that you update with += inside the loop
# Hint: a day with no reading has fields[2] == "" — `continue` moves on to the next line

**Step 3.** Build a dictionary mapping each month to the number of observations in it.

You do not know in advance which months are present, so the dictionary has to grow as you go.
Keep skipping the days with no reading: a day the sensor did not record is not an observation, so
October and November should come out short of their full length.

In [327]:
counts={} 

In [328]:
for line in lines:
    
    fields = line.strip().split(",")

    date = fields[1]

    months = date.split(".")[1]

    counts[months] = counts.get(months, 0) + 1


In [339]:
counts

{'01': 93,
 '02': 84,
 '03': 93,
 '04': 90,
 '05': 93,
 '06': 90,
 '07': 93,
 '08': 93,
 '09': 90,
 '10': 79,
 '11': 72,
 '12': 93}

In [340]:
# Your solution here
# Hint: start with an empty dict, counts = {}
# Hint: counts.get(month, 0) returns 0 for a month you have not seen yet
# Hint: counts[month] = counts.get(month, 0) + 1

**Step 4.** Build a second dictionary mapping each month to its **mean** daily maximum
temperature in °C. Print it with each mean rounded to two decimals.

In [341]:
monthly_totals={}

Extending the same loop below


In [342]:
for line in lines:
    fields = line.strip().split(",")
    if len(fields) != 3:
        skipped += 1
        continue

    _, date, temp_str = fields
    if temp_str == "":
        skipped += 1
        continue

    try:
        temp_f = float(temp_str)
    except ValueError:
        skipped += 1
        continue

    temp_c = fahrenheit_to_celsius(temp_f)
    month = date.split(".")[1]
    counts[month] = counts.get(month, 0) + 1
    monthly_totals[month] = monthly_totals.get(month, 0) + temp_c


In [ ]:
# Your solution here
# Hint: accumulate a running total per month in the same loop as the counts
# Hint: a second loop over the totals turns each total into a mean
# Hint: reuse your function from step 1 — do not repeat the conversion arithmetic

**Step 5.** Find the warmest month and print it with its mean temperature.

Do this with a loop and a comparison, keeping track of the best value seen so far, as you did for
the warmest day in Exercise 6. A dictionary is looked up by key, not by position, so there is no
`[0]` to start from: start from `None` instead.

In [347]:
monthly_means= {}

warmest_month = None
warmest_mean = None

In [349]:
#From previous loop. 

for line in lines:
    fields = line.strip().split(",")
    if len(fields) != 3:
        skipped += 1
        continue

    _, date, temp_str = fields
    if temp_str == "":
        skipped += 1
        continue

    try:
        temp_f = float(temp_str)
    except ValueError:
        skipped += 1
        continue

    temp_c = fahrenheit_to_celsius(temp_f)
    month = date.split(".")[1]
    counts[month] = counts.get(month, 0) + 1
    monthly_totals[month] = monthly_totals.get(month, 0) + temp_c

    monthly_means[month] = monthly_totals[month] / counts[month]


    for m, mean in monthly_means.items():
        if warmest_mean is None or mean > warmest_mean:
            warmest_month = m
            warmest_mean = mean

In [350]:
print(monthly_totals)
print(monthly_means)
print(warmest_mean)
print(warmest_month)

{'01': -675.5555555555552, '02': -30.833333333333364, '03': 1257.499999999999, '04': 1732.5, '05': 3443.333333333333, '06': 4117.222222222224, '07': 4336.11111111111, '08': 4135.555555555556, '09': 3563.333333333333, '10': 2130.8333333333326, '11': 461.9444444444443, '12': -215.83333333333331}
{'01': -2.7240143369175613, '02': -0.13764880952380967, '03': 5.0705645161290285, '04': 7.21875, '05': 13.884408602150536, '06': 17.155092592592602, '07': 17.48431899641577, '08': 16.67562724014337, '09': 14.847222222222221, '10': 10.707705192629811, '11': 2.6098556183301937, '12': -0.8702956989247311}
17.48431899641577
07


In [ ]:
# Your solution here
# Hint: initialise warmest_month = None and warmest_mean = None before the loop
# Hint: if warmest_mean is None, take this month; elif its mean is larger, take it instead

**Step 6.** Write your monthly summary to `_files/monthly_means.csv`, with the header
`month,n_obs,mean_temp_celsius` and one line per month, creating the `_files` folder first as you
did in 1.1. Then read the file back and print it, to confirm it says what you meant it to say.

In [ ]:
# Your solution here
# Hint: out_path = Path("_files") / "monthly_means.csv" builds the path, and
# out_path.parent.mkdir(exist_ok=True) creates the _files folder if it is missing
# Hint: open with mode "w" inside a with block, and write the header first
# Hint: build each line with an f-string, and remember the "\n"
# Hint: f"{value:.2f}" keeps the file tidy

**Step 7 (optional).** This step uses `try`/`except` from the going-deeper box in the lecture. Real files contain junk, and you have already met one kind of it — the empty
temperature field. Suppose some lines are truncated as well, or hold a temperature that is not a
number at all. Modify your reading loop so that it skips any line that does not have three fields
or whose temperature will not convert to a number, counting how many lines it skipped.

Test it on this list before applying it to the real file:

```python
test_lines = [
    "IIB,01.01.22,30.2",
    "IIB,02.01.22",
    "IIB,03.01.22,notanumber",
    "IIB,04.01.22,33.4",
]
```

You should keep 2 and skip 2.

In [ ]:
# Your solution here
# Hint: len(fields) != 3 catches the truncated line — use continue
# Hint: wrap the float() call in try/except ValueError to catch the bad number
# Hint: continue inside a for loop moves straight to the next line

**ℹ️ What you just did**

You wrote a reusable function, looped over a whole year of records, built dictionaries whose
keys you did not know in advance, aggregated by group, and wrote a derived dataset back to disk
— with a loop that survives malformed input. This is the shape of nearly every data-processing
script you will write. In subchapter 1.5, `pandas` will do the grouping in one line; the
reasoning will be exactly the same.

## Exercise 17: The solar system

<img src="https://raw.githubusercontent.com/gse-unil/2026_MLEES_book/main/part-I/_static/nasa-solar-system-montage.jpg" alt="The eight planets and the Moon photographed by spacecraft, arranged diagonally against black" width="100%">

<em>The eight planets and the Moon, as photographed by Mariner 10, Magellan, Galileo, Mars Global
Surveyor, Cassini and the Voyager spacecraft. The inner bodies are roughly to scale with each other, as are the outer ones, but the
two groups are not to scale with one another. Image
<a href="https://www.jpl.nasa.gov/images/pia03153-solar-system-montage-high-resolution-2001-version/">NASA/JPL (PIA03153)</a>, public domain.</em>


This is the long exercise for this subchapter, and the only one in it whose data is not
environmental. Eight planets and their masses are few enough to type by hand, and familiar enough
that a wrong answer is easy to spot, which makes them a good first test of lists, dictionaries and
functions together.

Work through sections A to C in order: section C writes functions against the dictionary you
build in section B. You will pick the exercise up again in 1.7, once classes are available, and
turn the dictionary into a `Planet` type.

The masses come from NASA's
[planetary fact sheet](https://nssdc.gsfc.nasa.gov/planetary/factsheet/). Use units of
10^24 kg throughout, so that Earth is 5.97 and Jupiter is 1898 — the variable names below carry
that unit, which is what makes a bare number like `1898` readable three cells later.

### A: lists and loops

**Q1) Create a list with the names of every planet in the solar system, in order.**

Name it `planets` and use lowercase names, so it starts `["mercury", "venus", ...]`.

In [ ]:
# Q1: create your list here

In [403]:
planets = ["mercury", "venus", "earth", "mars", "jupiter", "saturn", "uranus", "neptune"]


**Q2) Have Python tell you how many planets there are by examining your list.**

You should get 8. Do not type the number — read it off the list.

In [404]:
# Q2: count the planets
# Hint: len() reads the length off the list itself

In [405]:
print(len(planets))

8


**Q3) Use slicing to display the first four planets — the rocky ones.**

In [406]:
# Q3: the rocky planets

In [407]:
print(planets[:4])

['mercury', 'venus', 'earth', 'mars']


**Q4) Iterate through your planets and print the planet name only if it ends in "s".**

You should see `venus`, `mars` and `uranus`, and nothing else.

Hint: the last letter of a name is `planet[-1]`. Use `continue` for the names that do not match,
so the loop body says explicitly what it skips.

In [408]:
# Q4: print the planets whose name ends in s

In [409]:
for planet in planets:
    if planet[-1]== "s":
        continue
    print(planet)

mercury
earth
jupiter
saturn
neptune


### B: dictionary

**Q5) Create a dictionary that maps each planet name to its mass.**

Call it `planet_masses_1e24kg`, and read the values off the
[NASA fact sheet](https://nssdc.gsfc.nasa.gov/planetary/factsheet/) in units of 10^24 kg. Every
value has to be in that one unit: entered as 1.898 (in 10^27 kg), Jupiter would come out 1000
times too light in every comparison and ratio below.

In [410]:
# Q5: build the dictionary here

In [411]:
masses = {
    "mercury": 0.33,
    "venus": 4.87,
    "earth": 5.97,
    "mars": 0.642,
    "jupiter": 1898,
    "saturn": 568,
    "uranus": 86,
    "neptune": 102,
}

**Q6) Use your dictionary to look up Earth's mass.**

In [412]:
# Q6: look up Earth

print(masses["earth"])

5.97


**Q7) Loop through the dictionary and build a list of every planet heavier than
100 x 10^24 kg.**

You should end up with `["jupiter", "saturn", "neptune"]` — uranus, at 86.8, does not make it.

Hint: `.items()` gives you the name and the mass together on each pass, as in 1.2's lecture.

In [413]:
# Q7: the planets heavier than 100e24 kg


for planet,mass in masses.items():

    if mass > 100:
        print(planet)

jupiter
saturn
neptune


**Q8) Add Pluto to your dictionary.**

Its mass is 0.0130 x 10^24 kg. Print the names afterwards, looping with `.items()`, to confirm it
landed.

Pluto has not been a planet since 2006. Adding it still takes one assignment, and none of the code
that reads the other eight entries has to change.

In [414]:
# Q8: add Pluto, then print the names


masses["pluto"] = 0.013

#Confirming

for planet in masses.keys():
    print(planet)


mercury
venus
earth
mars
jupiter
saturn
uranus
neptune
pluto


### C: functions

**Q9) Write a function that converts a planetary mass to Earth masses.**

Jupiter is about 1898 x 10^24 kg, which is about 318 Earth masses:

$$M_{\text{Jupiter}} \approx 1898 \times 10^{24}\,\text{kg} \approx 318\,M_{\text{Earth}}$$

Call it `to_earth_masses(mass_1e24kg)` and check it against Jupiter — you should get
317.92294807370183.

In [415]:
# Q9: write to_earth_masses here, then test it on Jupiter


def earth_mass(planet):
    return masses[planet] / masses["earth"] 

In [416]:
earth_mass("jupiter")

317.92294807370183

**Q10) Now write a single function that converts to either Earth masses or Jupiter
masses, depending on a keyword argument.**

Call it `to_planet_masses(mass_1e24kg, reference="jupiter")`. Converting Jupiter's own mass should
give 317.923 with `reference="earth"` and exactly 1.0 with `reference="jupiter"`.

With the default argument, one function covers both conversions. Jupiter in Jupiter masses has
to come out as exactly 1.0; any other value means the function divides by the wrong mass.

In [417]:
# Q10: write to_planet_masses here, then test both references on Jupiter

In [418]:
def to_planet_mass(planet, reference = "earth"):
    if reference not in ["earth", "jupiter"]:
        raise ValueError("Ref mass must be 'earth' or 'jupiter'")
    return masses[planet] / masses[reference]

In [419]:
print(to_planet_mass("earth", reference="jupiter"))

print(to_planet_mass("jupiter", reference="earth"))

0.0031454162276080082
317.92294807370183


**Q11) Write a function that returns two values.**

`mass_in_two_units(mass_1e24kg, reference)` should return the mass in Earth masses *and* in masses
of whatever reference planet you pass. Test it on Jupiter with `reference="mars"`.

Hint: `return a, b` returns a tuple, and `first, second = f(...)` unpacks it — both from 1.2's
lecture.

In [420]:
# Q11: write mass_in_two_units here, then test it on Jupiter against mars

In [421]:
def mass_in_two_units(planet, reference1="earth", reference2="mars"):

    in_earth_mass= masses[planet]/masses[reference1]
    in_planet_b_mass= masses[planet]/masses[reference2]
    return in_earth_mass, in_planet_b_mass

In [422]:
jupiter_in_units = mass_in_two_units("jupiter", reference1="earth", reference2="mars")
print(jupiter_in_units)

(317.92294807370183, 2956.386292834891)


**Bonus.** Convert Neptune's mass to Jupiter masses with your Q10 function. Then pass that result
to your Q11 function with `reference="jupiter"`, and compare the second value it returns with the
Q10 result. Are they the same? If not, work out which function is being given a number in the
wrong unit, and what it should have been given instead — this is the mistake the `_1e24kg` suffix
exists to prevent.

In [ ]:
# Bonus: Neptune, out and back

In [424]:
neptune_into_jupiter = to_planet_mass("neptune", reference="jupiter")

In [425]:
print(neptune_into_jupiter)

0.05374077976817703


In [428]:
mass_neptune_two_units = mass_in_two_units("neptune", reference1="earth", reference2="jupiter")
print(mass_neptune_two_units)

(17.085427135678394, 0.05374077976817703)


**ℹ️ What you just did**

You built a list and read its length rather than counting it, sliced a subset out of it, and
looped over it with a condition. You built a dictionary keyed by name, looked a value up, filtered
it into a new list, and extended it. Then you wrote three functions over that dictionary: one
fixed, one with a default argument that collapses two functions into one, and one returning a pair.

That is the whole of this subchapter applied to one dataset. In 1.7 you will come back and replace
the dictionary with a class, so that a planet's name and mass travel together as one object
instead of as a key and a value that happen to line up.